service/history/mutableStateBuilder.go

```diff
@@ -3629,6 +3490,272 @@ func (e *mutableStateBuilder) GetContinueAsNew() *persistence.WorkflowSnapshot {
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if len(workflowEventsSeq) > 1 {
+		return nil, nil, &workflow.InternalServiceError{
+			Message: "cannot generate workflow snapshot with transient events",
+		}
+	}
+	if len(e.bufferedEvents) > 0 {
+		// TODO do we need the functionality to generate snapshot with buffered events?
+		return nil, nil, &workflow.InternalServiceError{
+			Message: "cannot generate workflow snapshot with buffered events",
+		}
+	}
+
+	if len(workflowEventsSeq) > 0 && e.GetReplicationState() != nil {
+		lastEvents := workflowEventsSeq[len(workflowEventsSeq)-1].Events
+		lastEvent := lastEvents[len(lastEvents)-1]
+		e.UpdateReplicationStateLastEventID(lastEvent.GetVersion(), lastEvent.GetEventId())
+	}
+
+	setTaskInfo(e.GetCurrentVersion(), now, e.pendingTransferTasks, e.pendingTimerTasks)
+
+	workflowSnapshot := &persistence.WorkflowSnapshot{
+		ExecutionInfo:    e.executionInfo,
+		ReplicationState: e.replicationState,
+
+		ActivityInfos:       convertPendingActivityInfos(e.pendingActivityInfoIDs),
+		TimerInfos:          convertPendingTimerInfos(e.pendingTimerInfoIDs),
+		ChildExecutionInfos: convertPendingChildExecutionInfos(e.pendingChildExecutionInfoIDs),
+		RequestCancelInfos:  convertPendingRequestCancelInfos(e.pendingRequestCancelInfoIDs),
+		SignalInfos:         convertPendingSignalInfos(e.pendingSignalInfoIDs),
+		SignalRequestedIDs:  convertSignalRequestedIDs(e.pendingSignalRequestedIDs),
+
+		TransferTasks:    e.pendingTransferTasks,
+		ReplicationTasks: e.pendingReplicationTasks,
+		TimerTasks:       e.pendingTimerTasks,
+
+		Condition: e.condition,
+	}
+
+	if err := e.closeSession(); err != nil {
+		return nil, nil, err
+	}
+	return workflowSnapshot, workflowEventsSeq, nil
+}
+
+func (e *mutableStateBuilder) closeSession() error {
+	// Clear all updates to prepare for the next session
+	e.hBuilder = newHistoryBuilder(e, e.logger)
+	e.updateActivityInfos = make(map[*persistence.ActivityInfo]struct{})
+	e.deleteActivityInfos = make(map[int64]struct{})
+	e.syncActivityTasks = make(map[int64]struct{})
+	e.updateTimerInfos = make(map[*persistence.TimerInfo]struct{})
+	e.deleteTimerInfos = make(map[string]struct{})
+	e.updateChildExecutionInfos = make(map[*persistence.ChildExecutionInfo]struct{})
+	e.deleteChildExecutionInfo = nil
+	e.updateRequestCancelInfos = make(map[*persistence.RequestCancelInfo]struct{})
+	e.deleteRequestCancelInfo = nil
+	e.updateSignalInfos = make(map[*persistence.SignalInfo]struct{})
+	e.deleteSignalInfo = nil
+	e.updateSignalRequestedIDs = make(map[string]struct{})
+	e.deleteSignalRequestedID = ""
+	e.continueAsNew = nil
+	e.continueAsNewWorkflowEvents = nil
+	e.clearBufferedEvents = false
+	if e.updateBufferedEvents != nil {
+		e.bufferedEvents = append(e.bufferedEvents, e.updateBufferedEvents...)
+		e.updateBufferedEvents = nil
+	}
+	if len(e.bufferedEvents) > e.config.MaximumBufferedEventsBatch() {
+		return ErrBufferedEventsLimitExceeded
```

Comment:
what does it mean when you return error at this stage ? What is the application supposed to do ? Can we return this error much earlier before we even flush events etc i.e. check if len(buffered) + len(updateBuffered) > limit ?

Reply:
right now this is handled by failing the decision
```
c.failInflightDecision()
```
ref: https://github.com/uber/cadence/blob/master/service/history/workflowExecutionContext.go#L750

The refactor need to incrementally change related logic